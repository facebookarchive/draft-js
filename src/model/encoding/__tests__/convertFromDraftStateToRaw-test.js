/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 * @oncall draft_js
 */

'use strict';

const BlockMapBuilder = require('BlockMapBuilder');
const CharacterMetadata = require('CharacterMetadata');
const ContentBlock = require('ContentBlock');
const ContentBlockNode = require('ContentBlockNode');
const ContentState = require('ContentState');
const DraftEntityInstance = require('DraftEntityInstance');

const convertFromDraftStateToRaw = require('convertFromDraftStateToRaw');
const getSampleStateForTesting = require('getSampleStateForTesting');
const Immutable = require('immutable');
const mockUUID = require('mockUUID');

jest.mock('uuid', () => jest.fn(mockUUID));

const {contentState} = getSampleStateForTesting();

const treeContentState = contentState.setBlockMap(
  BlockMapBuilder.createFromArray([
    new ContentBlockNode({
      key: 'A',
      children: Immutable.List.of('B', 'E'),
    }),
    new ContentBlockNode({
      parent: 'A',
      key: 'B',
      nextSibling: 'C',
      children: Immutable.List.of('C', 'D'),
    }),
    new ContentBlockNode({
      parent: 'B',
      key: 'C',
      text: 'left block',
      nextSibling: 'D',
    }),
    new ContentBlockNode({
      parent: 'B',
      key: 'D',
      text: 'right block',
      prevSibling: 'C',
    }),
    new ContentBlockNode({
      parent: 'A',
      key: 'E',
      text: 'This is a tree based document!',
      type: 'header-one',
      prevSibling: 'B',
    }),
  ]),
);

const getMetadata = (
  entityKey:
    | $TEMPORARY$string<'3'>
    | $TEMPORARY$string<'4'>
    | $TEMPORARY$string<'5'>
    | $TEMPORARY$string<'6'>,
) => Immutable.Repeat(CharacterMetadata.create({entity: entityKey}), 5);
const getLink = (
  entityKey:
    | $TEMPORARY$string<'3'>
    | $TEMPORARY$string<'4'>
    | $TEMPORARY$string<'5'>
    | $TEMPORARY$string<'6'>,
) =>
  new DraftEntityInstance({
    type: 'LINK',
    mutabiltity: 'MUTABLE',
    data: {
      url: `www.${entityKey}.com`,
    },
  });
// We start numbering our entities with '3' because getSampleStateForTesting
// already created an entity with key '2'.
const contentStateWithNonContiguousEntities = ContentState.createFromBlockArray(
  [
    new ContentBlock({
      key: 'a',
      type: 'unstyled',
      text: 'link2 link2 link3',
      characterList: getMetadata('3')
        .toList()
        .push(CharacterMetadata.EMPTY)
        .concat(getMetadata('4'))
        .push(CharacterMetadata.EMPTY)
        .concat(getMetadata('5')),
    }),
    new ContentBlock({
      key: 'b',
      type: 'unstyled',
      text: 'link4 link2 link5',
      characterList: getMetadata('5')
        .toList()
        .push(CharacterMetadata.EMPTY)
        .concat(getMetadata('3'))
        .push(CharacterMetadata.EMPTY)
        .concat(getMetadata('6')),
    }),
  ],
)
  .addEntity(getLink('3'))
  .addEntity(getLink('4'))
  .addEntity(getLink('5'))
  .addEntity(getLink('6'));

const assertConvertFromDraftStateToRaw = (content: ContentState) => {
  expect(convertFromDraftStateToRaw(content)).toMatchSnapshot();
};

test('must be able to convert from draft state with ContentBlock to raw', () => {
  assertConvertFromDraftStateToRaw(contentState);
});

test('must be able to convert from draft state with ContentBlockNode to raw', () => {
  assertConvertFromDraftStateToRaw(treeContentState);
});

test('must be able to convert from draft state with noncontiguous entities to raw', () => {
  assertConvertFromDraftStateToRaw(contentStateWithNonContiguousEntities);
});
