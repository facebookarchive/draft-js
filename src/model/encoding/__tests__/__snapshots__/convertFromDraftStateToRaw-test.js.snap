// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must be able to convert from draft state with ContentBlock to raw 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "a",
      "text": "Alpha",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 0,
          "length": 5,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [
        Object {
          "length": 5,
          "offset": 0,
          "style": "BOLD",
        },
      ],
      "key": "b",
      "text": "Bravo",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "c",
      "text": "Test",
      "type": "code-block",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "d",
      "text": "",
      "type": "code-block",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "e",
      "text": "",
      "type": "code-block",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [
        Object {
          "length": 7,
          "offset": 0,
          "style": "ITALIC",
        },
      ],
      "key": "f",
      "text": "Charlie",
      "type": "blockquote",
    },
  ],
  "entityMap": Object {
    "0": Object {
      "data": Object {},
      "mutability": "IMMUTABLE",
      "type": "IMAGE",
    },
  },
}
`;

exports[`must be able to convert from draft state with ContentBlockNode to raw 1`] = `
Object {
  "blocks": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "data": Object {},
              "depth": 0,
              "entityRanges": Array [],
              "inlineStyleRanges": Array [],
              "key": "C",
              "text": "left block",
              "type": "unstyled",
            },
            Object {
              "children": Array [],
              "data": Object {},
              "depth": 0,
              "entityRanges": Array [],
              "inlineStyleRanges": Array [],
              "key": "D",
              "text": "right block",
              "type": "unstyled",
            },
          ],
          "data": Object {},
          "depth": 0,
          "entityRanges": Array [],
          "inlineStyleRanges": Array [],
          "key": "B",
          "text": "",
          "type": "unstyled",
        },
        Object {
          "children": Array [],
          "data": Object {},
          "depth": 0,
          "entityRanges": Array [],
          "inlineStyleRanges": Array [],
          "key": "E",
          "text": "This is a tree based document!",
          "type": "header-one",
        },
      ],
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "A",
      "text": "",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {},
}
`;

exports[`must be able to convert from draft state with noncontiguous entities to raw 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 0,
          "length": 5,
          "offset": 0,
        },
        Object {
          "key": 1,
          "length": 5,
          "offset": 6,
        },
        Object {
          "key": 2,
          "length": 5,
          "offset": 12,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "a",
      "text": "link2 link2 link3",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 2,
          "length": 5,
          "offset": 0,
        },
        Object {
          "key": 0,
          "length": 5,
          "offset": 6,
        },
        Object {
          "key": 3,
          "length": 5,
          "offset": 12,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "b",
      "text": "link4 link2 link5",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {
    "0": Object {
      "data": Object {
        "url": "www.3.com",
      },
      "mutability": "IMMUTABLE",
      "type": "LINK",
    },
    "1": Object {
      "data": Object {
        "url": "www.4.com",
      },
      "mutability": "IMMUTABLE",
      "type": "LINK",
    },
    "2": Object {
      "data": Object {
        "url": "www.5.com",
      },
      "mutability": "IMMUTABLE",
      "type": "LINK",
    },
    "3": Object {
      "data": Object {
        "url": "www.6.com",
      },
      "mutability": "IMMUTABLE",
      "type": "LINK",
    },
  },
}
`;
